linux/ibus-keyman/src/engine.c

```diff
@@ -678,16 +707,59 @@ process_capslock_action(
   return TRUE;
 }
 
-static gboolean process_end_action(IBusKeymanEngine *keyman) {
-  keyman->firstsurrogate = 0;
-  if (keyman->char_buffer != NULL) {
-    ibus_keyman_engine_commit_string(keyman, keyman->char_buffer);
-    g_free(keyman->char_buffer);
-    keyman->char_buffer = NULL;
+static void
+commit_text(IBusKeymanEngine *keyman) {
+  g_assert(keyman != NULL);
+  if (keyman->commit_item > keyman->commit_queue) {
```

Comment:
Is this effectively an input condition? We could invert the test and return early and avoid nesting. (Do we want an assertion or log? I don't think so?)

Reply:
done